Add K8s.Conn.Auth.Exec authentication strategy

bforchhammer opened the pull request:
I've finally found some time to work on elixir-kubernetes things again, so here's a pull request for the exec auth module as previously discussed in #5, with some updates:

* added it as an option in the default list of authentication providers
* added some basic tests and updated some of the docs

I also tested it again with our EKS cluster and the `aws-iam-authenticator` (still working well!).

This is still missing support for `clientKeyData` and `clientCertificateData` -- is this something we should attempt to add or would it be okay to commit/merge without? Since I'm not sure how to test those I'd rather leave it for someone else...

Closes #5.

bforchhammer committed:
Add K8s.Conn.Auth.Exec authentication strategy

cjbottaro commented:
How do I use this? I tried updating my deps to use the repo/branch from this PR, but now it crashes when I try to get a connection:
```elixir
{:ok, conn} = K8s.Conn.lookup(:default)
** (FunctionClauseError) no function clause matching in K8s.Conn.Auth.Exec.format_env/1

    The following arguments were given to K8s.Conn.Auth.Exec.format_env/1:

        # 1
        nil

    Attempted function clauses (showing 2 out of 2):

        defp format_env(env) when is_list(env)
        defp format_env(%{"name" => key, "value" => value})

    (k8s 0.5.0) lib/k8s/conn/auth/exec.ex:61: K8s.Conn.Auth.Exec.format_env/1
    (k8s 0.5.0) lib/k8s/conn/auth/exec.ex:54: K8s.Conn.Auth.Exec.create/2
    (elixir 1.10.3) lib/enum.ex:3333: Enum.find_value_list/3
    (k8s 0.5.0) lib/k8s/conn.ex:135: K8s.Conn.from_file/2
    (k8s 0.5.0) lib/k8s/conn.ex:80: K8s.Conn.lookup/1
```
And here's my config:
```elixir
config :k8s, clusters: %{
  default: %{
    conn: kubeconfig,
    conn_opts: [context: "staging.eks.luna"]
  }
}
```

Thanks for the help.

cjbottaro commented:
Ahh, it looks like it's picky about kubeconfig. I had `env: null` in there. Changing to `env: []` makes it happy. Removing `env` completely results in a different crash.

Consider a clause:
```elixir
defp format_env(nil), do: %{}
```
?

coryodaniel commented:
> This is still missing support for clientKeyData and clientCertificateData -- is this something we should attempt to add or would it be okay to commit/merge without?

I think its fine to leave out if you dont need it. Id rather get some functionality for exec than getting it perfect on the first go!

bforchhammer commented:
Thanks for testing this @cjbottaro. Makes sense that it should work without env variables! I'll fix it and also rebase to `develop`...

bforchhammer committed:
Support env: null in kubeconfig

bforchhammer committed:
Make credo not fail because of TODO

bforchhammer commented:
* I've added some tests for `null` and missing env variables/ function arguments
* I changed credo to not fail if there's a `TODO` in code; @coryodaniel is that ok?
* I rebased the branch onto `develop` and resolved some minor merge conflicts

Please let me know if there's anything else to change?

coryodaniel commented:
I’d prefer to keep the TODO rule. Feel free to open an issue for the implementing the TODO though!

Thanks,

Cory O’Daniel

> On Jul 21, 2020, at 12:57 AM, Benedikt Forchhammer <notifications@github.com> wrote:
> 
> ﻿
> I've added some tests for null and missing env variables/ function arguments
> I changed credo to not fail if there's a TODO in code; @coryodaniel is that ok?
> I rebased the branch onto develop and resolved some minor merge conflicts
> Please let me know if there's anything else to change?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bforchhammer committed:
Removed ToDO

bforchhammer commented:
Ok, I removed the TODO and created https://github.com/coryodaniel/k8s/issues/85.

coryodaniel commented:
I appreciate that. I have a bad habit of leaving TOdO in code and that forces me to open tickets. 

I’ll cut this today!

Just curious how you found the docs/API around Auth. Was it a good experience, was anything bumpy, unclear, or weird?
